### PR TITLE
Fix error check for wrapped errors for safepath

### PR DESCRIPTION
**Pull request by alicefr**
**What this PR does / why we need it**:
Fix error checking with wrapped errors introduced by https://github.com/kubevirt/kubevirt/pull/8412

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-handler/container-disk/mount.go [hunk 1]
```diff
@@ -354,7 +355,7 @@ func (m *mounter) Unmount(vmi *v1.VirtualMachineInstance) error {
 		log.DefaultLogger().Object(vmi).Infof("Looking to see if containerdisk is mounted at path %s", entry.TargetFile)
 		file, err := safepath.NewFileNoFollow(entry.TargetFile)
 		if err != nil {
-			if os.IsNotExist(err) {
+			if errors.Is(err, os.ErrNotExist) {
```
Comment (vasiliy-ul) [inline pkg/virt-handler/container-disk/mount.go:358]: There is one more such `os.IsNotExist(err)` check on the line `476` in this file:

https://github.com/kubevirt/kubevirt/blob/b862288214babf12b06ce808efaf2adbf592e4a1/pkg/virt-handler/container-disk/mount.go#L476

Probably should also be replaced with `errors.Is(err, os.ErrNotExist)`

**Comment by vasiliy-ul** [comment]
Interesting finding :+1: I wasn't aware about the different behavior of `os.IsNotExist` vs `errors.Is(err, os.ErrNotExist)`. I grep'ed the sources and found a couple of more such checks:

https://github.com/kubevirt/kubevirt/blob/b862288214babf12b06ce808efaf2adbf592e4a1/pkg/virt-handler/non-root.go#L163-L175

and also in `pkg/network/setup/podnic.go`. But `podnic.go` does not seem to cause troubles (not 100% sure, though) and its probably not in the scope of this PR...

**Comment by alicefr** [comment]
Let me have another scan of the code. We have a mixture of the "standard" os library functions like os.Stat and function from the safepath library.

**Comment by alicefr** [comment]
I wanted to check first if the functional test mentioned by @iholder-redhat in https://github.com/kubevirt/kubevirt/pull/8464 was solved

**Comment by iholder-redhat** [comment]
Thanks a lot @alicefr! Great stuff!

For the record, I'd like to provide some context on these findings: some background and implications on the future, to reflect my offline conversations with @alicefr.

### Background
So first of all, kernel boot tests were quarantined [1][2]. I've tried to figure out what's the root cause and saw "file doesn't exist" failures regarding a socket path. Then I've look at the PRs that got merged right were the flakieness started, and noticed https://github.com/kubevirt/kubevirt/pull/8412. It raised my suspicion that some unwrapped error was returned somewhere, breaking the wrap-based `errors.Is()` check.

Then, I've created https://github.com/kubevirt/kubevirt/pull/8464 which basically reverts https://github.com/kubevirt/kubevirt/pull/8412 to check if this is indeed the issue. All lanes have passed twice in a row with no errors.

[1] https://github.com/kubevirt/kubevirt/pull/8459
[2] https://github.com/kubevirt/kubevirt/pull/8428

### Future implications
Error wrapping is a great language feature, but is a very delicate feature that has to be used with a lot of caution. The main problem is that it's very easy to destroy all wrapped data (e.g. by returning an unwrapped error: `return Errorf("... %v", err)`). If we want to use error wrapping - and I think we do since it's providing a lot of value - we need to think on how to make sure bugs aren't being introduced. Should we test it somehow? Or simply define conventions / policies?

I'm going to open an issue about it later today, just wanted to keep this info here for some context.

**Comment by iholder-redhat** [comment]
BTW - @alicefr - can you make sure the tests from both https://github.com/kubevirt/kubevirt/pull/8459 and https://github.com/kubevirt/kubevirt/pull/8428 pass with this PR's changes?

**Comment by alicefr** [comment]
> BTW - @alicefr - can you make sure the tests from both #8459 and #8428 pass with this PR's changes?
> 
> And a quick question: @alicefr do you know why the CI passed on your original PR? Was the `os.IsNotExist` introduced after your PR has merged?

I haven't been able to reproduce locally, that's why I want to check here if they still fail. I've checked the failed jobs you mentioned to me: 
  -  https://gcsweb.ci.kubevirt.io/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8445/pull-kubevirt-e2e-k8s-1.24-sig-compute/1569574972559462400/artifacts/k8s-reporter/1/
  - https://gcsweb.ci.kubevirt.io/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8449/pull-kubevirt-e2e-k8s-1.22-sig-compute/1569476121727078400/artifacts/k8s-reporter/3/

and the error messages are coming from the wrong check of the error type. I've encountered a similar issue during https://github.com/kubevirt/kubevirt/pull/8412. Probably, the rest of the errors aren't properly tested and/or flaky.

**Comment by alicefr** [comment]
I'm wondering if it would make sense to change all the `os.IsNotExist(err)` to `errors.Is(err, os.ErrNotExist)` in the rest of the code too. @vasiliy-ul @iholder-redhat wdyt?

**Comment by vasiliy-ul** [comment]
> I'm wondering if it would make sense to change all the `os.IsNotExist(err)` to `errors.Is(err, os.ErrNotExist)` in the rest of the code too. @vasiliy-ul @iholder-redhat wdyt?

+1. I think it does make sense: https://pkg.go.dev/os#IsNotExist

> It only supports errors **returned by the os package**. New code should use errors.Is(err, fs.ErrNotExist).

**Comment by iholder-redhat** [comment]
> > I'm wondering if it would make sense to change all the `os.IsNotExist(err)` to `errors.Is(err, os.ErrNotExist)` in the rest of the code too. @vasiliy-ul @iholder-redhat wdyt?
> 
> +1. I think it does make sense: https://pkg.go.dev/os#IsNotExist

Another +1 from me.
Thanks @vasiliy-ul for providing that link!

> 
> > It only supports errors **returned by the os package**. New code should use errors.Is(err, fs.ErrNotExist).

**Comment by rmohr** [comment]
> Fix error checking with wrapped errors introduced by #8412

Ha :(

> I'm wondering if it would make sense to change all the `os.IsNotExist(err)` to `errors.Is(err, os.ErrNotExist)` in the rest of the code too. @vasiliy-ul @iholder-redhat wdyt?

+1

**Comment by rmohr** [comment]
Just checked, seems like we don't need a backport for this.

**Comment by alicefr** [comment]
> > Fix error checking with wrapped errors introduced by #8412
> 
> Ha :(
> 

It seemed an innocent change at that time :/

**Comment by rmohr** [comment]
> > > Fix error checking with wrapped errors introduced by #8412
> > 
> > 
> > Ha :(
> 
> It seemed an innocent change at that time :/

Indeed :D

**Comment by rmohr** [comment]
Looks pretty good to me. One last idea: Would you mind adding something like

```diff
diff --git a/tools/analyzers/banncheck/banncheck.json b/tools/analyzers/banncheck/banncheck.json
index 077cc8a66..d320dbf4b 100644
--- a/tools/analyzers/banncheck/banncheck.json
+++ b/tools/analyzers/banncheck/banncheck.json
@@ -1,4 +1,10 @@
 {
+  "functions": [
+         {
+                 "name": "os.IsNotExist",
+                 "msg": "use errors.Is(err, os.ErrNotExist) instead of os.IsNotExist(err)"
+         }
+  ],
   "imports": [
     {
       "name": "kubevirt.io/kubevirt/pkg/unsafepath",
@@ -35,4 +41,4 @@
       ]
     }
   ]
-}
\ No newline at end of file
+}
```

That would tell people when they are compiling the code that they should not use os.IsNotExist.

**Comment by alicefr** [comment]
@rmohr can you suggest a good doc for the nogo analyzer? I want to exclude the vendor directory from the check, and I'm struggling a bit to find the right syntax.

**Comment by rmohr** [comment]
> @rmohr can you suggest a good doc for the nogo analyzer? I want to exclude the vendor directory from the check, and I'm struggling a bit to find the right syntax.

It should already be excluded by using `only_files` in `nogo_config.json`:

```
  "bannedAPI": {
    "only_files": {
      "pkg/": "only production code must not use banned imports"
    },
    "exclude_files": {
      "pkg/.*_test\\.go": "unit tests are allowed to use banned imports"
    }
  },
```

But in principle you can just exclude the whole vendor directory, like on some other linters:

```
  "ginkgolinter": {
    "exclude_files": {
      "vendor/": "vendor doesn't pass vet",
      "external/": "externaldoesn't pass vet"
    }
  },
```

**Comment by alicefr** [comment]
@rmohr thanks :) this commit includes: https://github.com/kubevirt/kubevirt/pull/8492/commits/411874563c39a0b936d7967ea3ff323631d365e7 the linter fix

**Comment by iholder-redhat** [comment]
Thanks @rmohr! Enforcing it via a linter is a great step forward!
I just want to point out that while it's great we still need to think further on how to handle error wrapping in general. Using `error.Is()` will still fail to work if the returned error isn't being wrapped.

**Comment by alicefr** [comment]
> Thanks @rmohr! Enforcing it via a linter is a great step forward! I just want to point out that while it's great we still need to think further on how to handle error wrapping in general. Using `error.Is()` will still fail to work if the returned error isn't being wrapped.

I'm really not sure. This depends on how you decide to construct the error. The function `error.Is` allows to handle both cases:
  1) when the error is exactly `os.ErrNotExist`
  2) a wrapped version of `os.ErrNotExist`

but it is different for example than enforcing always using `ftm.Errorf("%w", error)` instead of `ftm.Errorf("%v", error)`. It could be still a good idea enforcing this practise, I think cri-o does it.

**Comment by rmohr** [comment]
> Thanks @rmohr! Enforcing it via a linter is a great step forward!
> I just want to point out that while it's great we still need to think further on how to handle error wrapping in general. Using `error.Is()` will still fail to work if the returned error isn't being wrapped.

In general I would want to use `%w` as little as possible, because it can create implicit dependencies across the whole stack, introducing hidden APIs. Therefore I am not a super-big fan of `%w`. I however think that checking for `os.IsNotExist` is such a common need that I am ok with it, since without being able to enriching it, while still detecting it, the info propagated is very often uncomfortably thin.

One possible solution would be to have an explicit allow-list of which errors are allowed to be checked for with `errors.Is`, but all this is a little bit tricky.

**Comment by iholder-redhat** [comment]
> but it is different for example than enforcing always using `fmt.Errorf("%w", error)` instead of `fmt.Errorf("%v", error)`. It could be still a good idea to enforce this practise, I think cri-o does it.

That would solve the problem, but it might introduce a different kind of problem which is exposing our internal error structure to users.

For example, let's say that we would decide to use error wrapping internally. So imagine we define errors such as `ErrVmCannotStart`, etc. If we expose these externally by returning errors wrapped with these, users could rely on the assumptions that we would always return the same wrapped errors (e.g. user code could do `if errors.Is(err, ErrVmCannotStart) {...`. This can make it harder to refactor later and may expose internal structures / flows that we don't want to expose.

I think that we shouldn't use a "black-or-white" solution, but instead think of how a policy / convention that we want to use to ensure that only errors that we want to expose are being wrapped and handed to the user. I think we should open an issue on that and discuss this further, but obviously it's outside of this PR's scope.

**Comment by rmohr** [comment]
/lgtm

looks good. Some e2e tests look a little bit strange. Let's see how they are doing before approving.

**Comment by iholder-redhat** [comment]
> > I think we should open an issue on that and discuss this further, but obviously it's outside of this PR's scope.
> 
> Creating an issue sounds good. Thanks. +1

This is a start: https://github.com/kubevirt/kubevirt/issues/8512 :)

**Pull request merged**